compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs

```diff
@@ -4436,6 +4436,43 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         }
     }
 
+    /// If the type failed selection but the trait is implemented for `(T,)`, suggest that the user
+    /// creates a unary tuple
+    ///
+    /// This is a common gotcha when using libraries that emulate variadic functions with traits for tuples.
+    pub(super) fn suggest_tuple_wrapping(
+        &self,
+        err: &mut Diag<'_>,
+        root_obligation: &PredicateObligation<'tcx>,
+        obligation: &PredicateObligation<'tcx>,
+        span: Span,
+    ) {
+        if !matches!(obligation.cause.code(), ObligationCauseCode::FunctionArg { .. }) {
+            return;
+        }
+
+        let Some(root_pred) = root_obligation.predicate.as_trait_clause() else { return };
+
+        let trait_ref = root_pred.map_bound(|root_pred| {
+            root_pred
+                .trait_ref
+                .with_self_ty(self.tcx, Ty::new_tup(self.tcx, &[root_pred.trait_ref.self_ty()]))
+        });
+
+        self.infcx.enter_forall(trait_ref, |trait_ref| {
+            if self
+                .type_implements_trait(trait_ref.def_id, trait_ref.args, root_obligation.param_env)
+                .must_apply_modulo_regions()
+            {
+                err.multipart_suggestion_verbose(
+                    format!("use a unary tuple instead"),
+                    vec![(span.shrink_to_lo(), "(".into()), (span.shrink_to_hi(), ",)".into())],
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        });
+    }
```

Comment:
this is kinda unnecessary. instead of using `enter_forall` and calling `type_implements_trait`, just use [`predicate_must_hold_modulo_regions`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/query/evaluate_obligation/trait.InferCtxtExt.html#tymethod.predicate_must_hold_modulo_regions).